lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1264 @@
+      Inst::Kind K = Inst::getKind(KindStr);
+      UserCompKinds.insert(K);
+      if (KindStr == Inst::getKindName(Inst::Const)) { // Special case
+        ConstComps.push_back(Component{Inst::Const, 0, {}});
+      } else if (K == Inst::ZExt || K == Inst::SExt || K == Inst::Trunc)
+        continue; // handled in addZSTComps
+      else if (K == Inst::Kind(~0))
+        report_fatal_error("unknown instruction: " + KindStr + "\n");
+      else if (UnsupportedCompKinds.count(K))
+        report_fatal_error("unsupported instruction: " + KindStr + "\n");
+      else
+        Kinds.push_back(K);
+    }
+    for (auto const &Comp : CompLibrary)
+      for (auto const &Kind : Kinds)
+        if (Comp.Kind == Kind)
+          Comps.push_back(Comp);
+  } else {
+    Comps = CompLibrary;
+    ConstComps.push_back(Component{Inst::Const, 0, {}});
+    llvm::outs() << "WARNING: using all " << Comps.size()
+                 << " components, synthesis will probably take long time"
+                 << " or run out of memory\n";
+  }
+}
+
+void InstSynthesis::initInputVars(Inst *LHS, InstContext &IC) {
+  std::vector<Inst *> Tmp;
+  getInputVars(LHS, Tmp);
+  // Remove duplicates
+  std::set<Inst *> TmpSet;
+  for (auto I : Tmp)
+    if (TmpSet.insert(I).second)
+      Inputs.push_back(I);
+  for (unsigned J = 0; J < Inputs.size(); ++J) {
+    // Note that location variable 0_0 is not used
+    LocVar In = std::make_pair(0, J+1);
+    std::string LocVarStr = getLocVarStr(In, LOC_PREFIX);
+    Inst *Loc = IC.createVar(LocInstWidth, LocVarStr);
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Add input location to I
+    I.emplace_back(In, Loc);
+    // Update input name
+    LocVarStr = getLocVarStr(In, INPUT_PREFIX);
+    Inputs[J]->Name = LocVarStr;
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Update CompInstMap map with concrete Inst
+    CompInstMap[In] = Inputs[J];
+  }
+}
+
+void InstSynthesis::initConstComponents(InstContext &IC) {
+  for (unsigned J = 0; J < ConstComps.size(); ++J) {
+    auto &Comp = ConstComps[J];
+    // Treat constant comps as ordinary inputs
+    LocVar In = std::make_pair(0, Inputs.size()+J+1);
+    std::string LocVarStr = getLocVarStr(In, LOC_PREFIX);
+    Inst *Loc = IC.createVar(LocInstWidth, LocVarStr);
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Add constant component location to I
+    I.emplace_back(In, Loc);
+    // Update name
+    LocVarStr = getLocVarStr(In, CONST_PREFIX);
+    LocInstMap[LocVarStr] = std::make_pair(In, Loc);
+    // Update CompInstMap map with concrete Inst
+    Loc = IC.createVar(DefaultWidth, LocVarStr);
+    CompInstMap[In] = Loc;
+  }
+}
+
+void InstSynthesis::setDefaultWidth(Inst *LHS) {
+  // Set the DefaultWidth of component instances to the max width seen in inputs
+  for (auto In : Inputs)
+    if (In->Width > DefaultWidth)
+      DefaultWidth = In->Width;
+  // No inputs, use output width
+  if (!DefaultWidth)
```

Comment:
Can you explain how no inputs can happen? Wouldn't any Souper expression have at least one input?